azuredevops/internal/utils/debug/Debug.go

```diff
@@ -0,0 +1,33 @@
+package debug
+
+import (
+	"log"
+	"os"
+	"strconv"
+	"time"
+)
+
+var debugWaitPassed = false
+
+const defaultWait = 20
+
+// Wait wait to attach a debugger if the environment variable AZDO_PROVIDER_DEBUG is set to 1
+// In addition the function supports reading the wait time from the AZDO_PROVIDER_DEBUG_WAIT environment variable
+// The default waiting time is 20 seconds
+func Wait(force ...bool) {
```

Comment:
Why is this needed?

Reply:
Personally I got fed up with adding this infamous `time.Sleep` to be able to attach the debugger to a running Go process. So I thought it could be worth adding a function that could be controlled via some environment variables. If we'd use this for every resource or data source, people could attach debuggers even for a official version of the provider, so that they could provide us with more detailed error reports or even pull requests at best.

Reply:
Why do we need time.sleep to attach the debugger? What have I missed?

Reply:
How are you debugging the provider?

Reply:
> Why do we need time.sleep to attach the debugger? What have I missed?

Mostly for debug purpose.  It 's some kind like bus station waiting for passengers. 
[IDEA debug go application](https://blog.jetbrains.com/go/2019/02/06/debugging-with-goland-getting-started/) roll a little down to `Debugging a running application on the local machine
`

Reply:
Hi @tmeckel, I prefer acceptance. As acceptance will create the real resource, when the acceptance test is executing, I can debug the provider source code. Combine with `TF_LOG=TRACE`, `TF_LOG_PATH=log_file_path` configures. This covers most scenarios for me.
I don't know what develop environment you are. I use Goland, it might different from you.

Reply:
From my point of it's very valuable to have an debugger attached to a running provider because you can analyze current values in variables what gives you often a hint why things go wrong and with wrong I mean, you don't get a crash dump (panic) but simply your code doesn't work as expected.

I think I mostly learned about Terraform and it's providers by stepping through using a debugger. Another advantage using a debugger is the fact that you don't rely on people adding (meaningful) log messages. My experience is that this is mostly not the case 😁 

And about @EliiseS question why the wait (sleep) is needed: Terraform starts/executes the providers as a new process. So to have the ability to attach to the "spawned" provider process you must have some kind of waiting mechanism, so that you have the time to extract the process id of the provider process and use VSCode (dlv) to attach to that process. Go doesn't have this luxury like in .Net or Windows Native applications where you can add an `Debugger.Launch()` which automatically attaches Visual Studio to the running process.

https://github.com/hashicorp/terraform/issues/20466
https://github.com/hashicorp/terraform/pull/16305

Reply:
Recommend debug with acceptance test during your development. Don't need additional code for debug purpose and we can debug directly from the source code not the binary file.  More smooth and slight.

Reply:
Okay ... so what's the decision here? Keep the code? Remove it?

Reply:
How are you debugging the provider? I'm unable to get debug to work for with the acceptance tests, but debug for unit tests I can just press the `debug test` button above the test in VSCode.

Reply:
Debugging unit tests won't give you any insights to the interaction between Terraform, the provider and the backend API that the provider is using, because unit tests shouldn't use any APIs aside the unit (module) which is under test. That's why @nmiodice (or someone other clever person) introduced the usage of API mocks for the DevOps provider. 

This is also the reason why @xuzhang3 favors the use of acceptance tests so much. The unit tests don't test all components together. As good and important as acceptance testing is, I believe that certain bugs cannot be solved by using acceptance testing alone, or the solution takes much longer. Therefore I use the GO Debugger to analyze and solve persistent problems and to follow the inner process. 

I always proceed the same way:

1. create a HCL file that calls up the exact area of the provider where the problem is.
2. add a `time.Sleep` method typically at a function that is called by Terraform like `read`, `update` etc.
3. compiling the provider with additional debug flags. That's why I once added those flags to `build.sh` and filed a PR 😁 
4. Call `terraform plan` or `apply`
5. determine the process ID of the provider that was started by Terraform
6. entering the process ID in the `launch.json` of VSCode
7. press F5 at VSCode to attach the debugger to the provider
8. perform debugging

How to debug Go with a debugger is described here: https://github.com/golang/vscode-go/blob/master/docs/debugging.md

Reply:
> How are you debugging the provider? I'm unable to get debug to work for with the acceptance tests, but debug for unit tests I can just press the `debug test` button above the test in VSCode.

`dlv` is a good tool for debugging [delve](https://github.com/go-delve/delve). 
Usage example (windows environment): 
- open CMD 
- cd to the test file directory
- `dlv test .  -- -test.run=TestAccExampleService_basic -test.v`
- enjoy your debug steps

Reply:
> Okay ... so what's the decision here? Keep the code? Remove it?

The debug help file should be removed. You can keep this help files in your person branch.   [dlv](https://github.com/go-delve/delve/) may be an optional for you for debug.

Reply:
[GDB](https://golang.org/doc/gdb) is also an optional for debugging

Reply:
Removed the code